packages/core/src/mm/actors/Player.c

```diff
@@ -1644,3 +1738,437 @@ void Player_UseItem(GameState_Play* play, Actor_Player* this, s16 itemId)
         Player_UseItemImpl(play, this, itemId);
     }
 }
+
+/* Hammer Stuff */
+
+s32 Player_CustomActionToModelGroup(Actor_Player* player, s32 itemAction) {
+    u8* sActionModelGroups = (u8*)0x801BFF3C; /* using original table also means original glitches, if that matters */
+    s32 modelGroup = sActionModelGroups[itemAction];
+
+    if (itemAction == PLAYER_CUSTOM_IA_HAMMER) {
+        modelGroup = sActionModelGroups[7]; /* uses deku stick model group but does not draw deku stick because of the way the original draw code for it works */
+    }
+
+    /* if ((modelGroup == PLAYER_MODELGROUP_ONE_HAND_SWORD) && Player_IsGoronOrDeku(player)) { */
+    if ((modelGroup == 2) && (player->transformation == MM_PLAYER_FORM_GORON || player->transformation == MM_PLAYER_FORM_DEKU)) {
+        /* return PLAYER_MODELGROUP_1;  */
+        return 1;
+    }
+    return modelGroup;
+}
+
+/*
+    first two arguments match Player_SetUpperAction which allows it to replace relevant calls in which the third argument would always be this->heldItemAction
+*/
+void Player_SetCustomItemActionUpperFunc(GameState_Play* play, Actor_Player* player) {
+    PlayerUpperActionFunc* sPlayerUpperActionUpdateFuncs = (PlayerUpperActionFunc*)OverlayAddr(0x8085c9f0);
+    void (*Player_SetUpperAction)(GameState_Play* play, Actor_Player* this, PlayerUpperActionFunc upperActionFunc) = OverlayAddr(0x8082f43c);
+
+    if (player->heldItemAction == PLAYER_CUSTOM_IA_HAMMER) {
+        Player_SetUpperAction(play, player, sPlayerUpperActionUpdateFuncs[6]);
+    } else {
+        Player_SetUpperAction(play, player, sPlayerUpperActionUpdateFuncs[player->heldItemAction]);
+    }
```

Comment:
Maybe it's just my style preference but i'd write it more like this.
```suggestion
    s8 heldItemAction = player->heldItemAction;
    if (heldItemAction == PLAYER_CUSTOM_IA_HAMMER) {
        heldItemAction = PLAYER_IA_SWORD_TWO_HANDED;
    }
    Player_SetUpperAction(play, player, sPlayerUpperActionUpdateFuncs[heldItemAction ]);
```

Reply:
If it's alright I would prefer not to include this change because setting a variable named `heldItemAction` to something that is not actually the held item action is misleading even if the functionality is identical. I also don't know if the compiler would just optimize it out but I wonder if this change would cause the 6 to be placed in a register instead of as an added offset in the LW instruction, which I dislike. I did replace the 6 with PLAYER_IA_SWORD_HAMMER to make it clearer why this exists though. If you guys are thinking about the OoT Boomerang it would be worth considering adding a `sPlayerCustomUpperActionUpdateFuncs` alternative.

Reply:
Kinda agree with @ZoeyZolotova on this one. I guess `itemAction` would be an okay name? Much better to have a single call IMO.

Also don't worry about the compiler too much. We must aggressively optimize for space anyway, this will give it more leverage.